py/obj.c

```diff
@@ -202,67 +202,93 @@ bool mp_obj_is_callable(mp_obj_t o_in) {
 // Furthermore, from the v3.4.2 code for object.c: "Practical amendments: If rich
 // comparison returns NotImplemented, == and != are decided by comparing the object
 // pointer."
-bool mp_obj_equal(mp_obj_t o1, mp_obj_t o2) {
-    // Float (and complex) NaN is never equal to anything, not even itself,
-    // so we must have a special check here to cover those cases.
-    if (o1 == o2
-        #if MICROPY_PY_BUILTINS_FLOAT
-        && !mp_obj_is_float(o1)
-        #endif
-        #if MICROPY_PY_BUILTINS_COMPLEX
-        && !mp_obj_is_type(o1, &mp_type_complex)
-        #endif
-        ) {
-        return true;
-    }
-    if (o1 == mp_const_none || o2 == mp_const_none) {
-        return false;
-    }
+mp_obj_t mp_obj_equal_bop(mp_binary_op_t op, mp_obj_t o1, mp_obj_t o2) {
+    mp_obj_t local_true = (op == MP_BINARY_OP_NOT_EQUAL) ? mp_const_false : mp_const_true;
+    mp_obj_t local_false = (op == MP_BINARY_OP_NOT_EQUAL) ? mp_const_true : mp_const_false;
+    int pass_number = 0;
 
-    // fast path for small ints
-    if (mp_obj_is_small_int(o1)) {
-        if (mp_obj_is_small_int(o2)) {
-            // both SMALL_INT, and not equal if we get here
-            return false;
-        } else {
-            mp_obj_t temp = o2; o2 = o1; o1 = temp;
-            // o2 is now the SMALL_INT, o1 is not
-            // fall through to generic op
-        }
+    // Shortcut for very common cases
+    if (o1 == o2 &&
+        (mp_obj_is_small_int(o1) ||
+	 ((mp_obj_get_type(o1)->flags & MP_TYPE_FLAG_NO_EQUALITY_SHORTCUTS) == 0) )) {
+        return local_true;
     }
 
+    bool is_str_1 = mp_obj_is_str(o1);
+    bool is_str_2 = mp_obj_is_str(o2);
+
     // fast path for strings
-    if (mp_obj_is_str(o1)) {
+    if (is_str_1 && is_str_2) {
```

Comment:
Is this correct?  I think it should be `||` not `&&`... anyway, I can take it from here, I think I can make this a bit simpler.

Reply:
There's definitely something wrong either here or the line below. Let me investigate and fix before you merge.

Reply:
I've already fixed it and have done a few other clean ups.... I can push it as a new PR for further review.